### ISSUE TITLE
Dependency on `default_data` broken

### ISSUE THREAD

**Issue by hdoupe** [open-source-economics/Tax-Calculator#2115]
@martinholmer I'm sorry I missed this when reviewing PR #2103. The webapp relies on the `start_year` argument in `Parameters.default_data`. The data returned from this method populates the placeholders on the inputs page. This way, the user can toggle the `start_year` and view the default values for the selected year.

Can we add the ability to specify a `start_year` in `default_data`? I was able to restore this capability by overriding the new `default_data` class method and adding back the `_revised_default_data` method:

```python
class TCParamsMixin:

    @classmethod
    def default_data(cls, metadata=False, start_year=None):    # same as tc 0.22.2 default_data except for start_year arg in cls()
        """
        Return parameter data read from the subclass's json file.
        Parameters
        ----------
        metadata: boolean
        start_year: int or None
        Returns
        -------
        params: dictionary of data
        """
        # extract different data from DEFAULT_FILENAME depending on start_year
        if start_year is None:
            params = cls._params_dict_from_json_file()
        else:
            nyrs = start_year - cls.JSON_START_YEAR + 1
            ppo = cls()                       # <--- only change, used to be ppo = cls(num_years=nyrs)
            ppo.set_year(start_year)
            params = getattr(ppo, '_vals')
            params = TCParamsMixin._revised_default_data(params, start_year,
                                                      nyrs, ppo)
        # return different data from params dict depending on metadata value
        if metadata:
            return params
        else:
            return {name: data['value'] for name, data in params.items()}

    @staticmethod
    def _revised_default_data(params, start_year, nyrs, ppo): # same as tc 0.22.2 _revised_default_data
        """
        Return revised default parameter data.
        Parameters
        ----------
        params: dictionary of NAME:DATA pairs for each parameter
            as defined in calling default_data staticmethod.
        start_year: int
            as defined in calling default_data staticmethod.
        nyrs: int
            as defined in calling default_data staticmethod.
        ppo: Policy object
            as defined in calling default_data staticmethod.
        Returns
        -------
        params: dictionary of revised parameter data
        Notes
        -----
        This staticmethod is called from default_data staticmethod in
        order to reduce the complexity of the default_data staticmethod.
        """
        start_year_str = '{}'.format(start_year)
        for name, data in params.items():
            data['start_year'] = start_year
            values = data['value']
            num_values = len(values)
            if num_values <= nyrs:
                # val should be the single start_year value
                rawval = getattr(ppo, name[1:])
                if isinstance(rawval, np.ndarray):
                    val = rawval.tolist()
                else:
                    val = rawval
                data['value'] = [val]
                data['row_label'] = [start_year_str]
            else:  # if num_values > nyrs
                # val should extend beyond the start_year value
                data['value'] = data['value'][(nyrs - 1):]
                data['row_label'] = data['row_label'][(nyrs - 1):]
        return params


class WebappPolicy(TCParamsMixin, taxcalc.Policy):
    pass

class WebappBehavior(TCParamsMixin, taxcalc.Behavior):
    pass
```

Then, I just called the `default_data` method with `WebappPolicy.default_data(start_year=start_year)` and `WebappBehavior.default_data(start_year=start_year)`

@martinholmer What do you think is the best way to go about resolving this issue? I would prefer not to add the above code to the webapp on a permanent basis. I also understand that keeping code in Tax-Calculator that is only used by the webapp is a burden on Tax-Calculator developers.

**Comment by martinholmer** [comment]
@hdoupe said in T-C issue #2115:
> I'm sorry I missed this when reviewing PR #2103. [TaxBrain] relies on the `start_year` argument in `Parameters.default_data`. The data returned from this method populates the placeholders on the inputs page. This way, the user can toggle the `start_year` and view the default values for the selected year.
>
> Can we add the ability to specify a `start_year` in `default_data`?

I'm not sure that is the best way to deal with the situation because you haven't supplied enough information.

1. Can you point us to the TaxBrain code that calls the Policy.default_data() method?

2. What about the other Tax-Calculator classes that are derived from the Parameters class; does TaxBrain ever call their `default_data` method?

**Comment by hdoupe** [comment]
@martinholmer asked:

> 1. Can you point us to the TaxBrain code that calls the Policy.default_data() method?

Yes, here is one example: https://github.com/ospc-org/ospc.org/blob/stable/webapp/apps/taxbrain/param_displayers.py#L201-L211

> 2. What about the other Tax-Calculator classes that are derived from the Parameters class; does TaxBrain ever call their default_data method?

TaxBrain also calls the `default_data` method on the `Behaviors` class. However, the behavior parameters don't really depend on the time dimension. So, this can be worked around.

**Comment by martinholmer** [comment]
@hdoupe said in Tax-Calculator issue #2115:
> here is one example [of where TaxBrain calls Policy.default_data with a start_year argument]: https://github.com/ospc-org/ospc.org/blob/stable/webapp/apps/taxbrain/param_displayers.py#L201-L211

OK.  But why not just use instead this TaxBrain `default_policy` function?
```
# Create a list of default policy
def default_policy(first_budget_year, use_puf_not_cps=True):

    TAXCALC_DEFAULT_PARAMS_JSON = default_taxcalc_data(taxcalc.policy.Policy,
                                                       metadata=True,
                                                       start_year=first_budget_year)

    default_taxcalc_params = {}
    for k,v in TAXCALC_DEFAULT_PARAMS_JSON.items():
        param = TaxCalcParam(k,v, first_budget_year,
                             use_puf_not_cps=use_puf_not_cps)
        default_taxcalc_params[param.nice_id] = param

    TAXCALC_DEFAULT_PARAMS = default_taxcalc_params

    return TAXCALC_DEFAULT_PARAMS
```

**Comment by martinholmer** [comment]
@hdoupe said in Tax-Calculator issue #2115:
> TaxBrain also calls the default_data method on the Behaviors class. However, the behavior parameters don't really depend on the time dimension. So, this can be worked around.

OK.  That's good to know.

**Comment by hdoupe** [comment]
@martinholmer That function calls [`default_taxcalc_data`](https://github.com/ospc-org/ospc.org/blob/stable/webapp/apps/taxbrain/helpers.py#L170-L181) which also uses the `default_data`method:

```python
def default_taxcalc_data(cls, start_year, metadata=False):
    ''' Call the default data function on the given class for the given
        start year with meatadata flag
    '''
    dd = cls.default_data(metadata=metadata, start_year=start_year)
    if metadata:
        for k in dd:
            dd[k]['value'] = round_gt_one_to_nearest_int(dd[k]['value'])
    else:
        for k in dd:
            dd[k] = round_gt_one_to_nearest_int(dd[k])
    return dd
```

**Comment by martinholmer** [comment]
@hdoupe in Tax-Calculator issue #2115 responded to this question as follows:
>> Can you point us to the TaxBrain code that calls the Policy.default_data() method?
>
> here is one example: https://github.com/ospc-org/ospc.org/blob/stable/webapp/apps/taxbrain/param_displayers.py#L201-L211

without mentioning where the other calls to Policy.default_data() were.

So, I've had to look at the webapp/apps/taxbrain portion of the source code.  It's pretty complicated.  But as best as I can tell, the calls to Policy.default_data() with a specified calendar year were a way to get two very different kinds of information:
(1)  Information (like sections) that does not change over time, and
(2)  Current-law parameter values for the specified calendar year.

@hdoupe, is that correct?  If so, it seems to me that there is no need for the old `default_data()` method that accepted a calendar year for the following reasons.  

When a policy object is instantiated (via `clp = Policy()`), the dictionary that is returned by the `default_data()` method is saved as `clp._vals` and, as far as I can see, is never modified.  So that means that the unchanging (1) information is available from the Policy object no matter what its `current_year` value is.  

And the changing (2) information is available by simply moving the current-law policy object to any specified calendar year (via `clp.advance_to_year(calyear)`).

Perhaps I'm missing something, but why wouldn't this approach work for TaxBrain?

**Comment by martinholmer** [comment]
This comment in Tax-Calculator issue #2115 present results of an investigation into the results returned by a call to `Policy.default_data(metadata=True, start_year=2017)` in Tax-Calculator release 0.22.2.  This is the static method that @hdoupe [wants to reinstitute](https://github.com/open-source-economics/Tax-Calculator/issues/2115#issue-381302377) in the next Tax-Calculator release.

What the investigation finds is that for many policy parameters the values of the `row_label` and `value` keys in the returned dictionary are very misleading.  The returned results are so misleading that they could easily cause errors when being used in other applications.  Given this level of confusion, I consider the old (release 0.22.2) results returned by a call to `Policy.default_data(metadata=True, start_year=2017)` to be buggy.

The problem with the old `Policy.default_data(metadata=True, start_year=2017)` results is that many parameters have `row_label` and `value` information for years **after** the specified `start_year`.  Here is an example:
```
_II_em
long_name Personal and dependent exemption amount
description Subtracted from AGI in the calculation of taxable income, per taxpayer and dependent.
section_1 Personal Exemptions
section_2 Personal And Dependent Exemption Amount
irs_ref Form 1040, line 42. 
notes 
row_var FLPDYR
row_label ['2017', '2018', '2019', '2020', '2021', '2022', '2023', '2024', '2025', '2026']
start_year 2017
cpi_inflatable True
cpi_inflated True
col_var 
col_label 
boolean_value False
integer_value False
value [4050.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4883.0]
range OrderedDict([('min', 0), ('max', 9e+99)])
out_of_range_minmsg 
out_of_range_maxmsg 
out_of_range_action stop
compatible_data OrderedDict([('puf', True), ('cps', True)])
```
It would be far less confusing, and hence reduce the chance of errors in applications calling this method, if the `row_label` value was `['2017']` and the `value` value was `[4050.0]`.   In fact, many other scalar policy parameters do have those values of length one in the returned dictionary.  So, there is no consistency in the pattern of returned results from a call to `Policy.default_data(metadata=True, start_year=2017)` in Tax-Calculator release 0.22.2.

@MattHJensen @hdoupe

**Comment by hdoupe** [comment]
@martinholmer said:
> So, I've had to look at the webapp/apps/taxbrain portion of the source code. It's pretty complicated. But as best as I can tell, the calls to Policy.default_data() with a specified calendar year were a way to get two very different kinds of information:
(1) Information (like sections) that does not change over time, and
(2) Current-law parameter values for the specified calendar year.
>
> @hdoupe, is that correct?


I'm sorry that you had to look at that code. It is complicated. I've spent the last two working days refactoring it so that it is more intuitive and understandable. Yes, reasons (1) and (2) are why `default_data` is used.

I agree with your assessment about the result of `default_data`. I'll have some feedback for you today on how things go with the new result structure.

**Comment by martinholmer** [comment]
Issue #2115 has been resolved by pull request #2119.
